### PR TITLE
Eval: enhance eval create skill with coverage distribution, multi-turn support, and quality rules

### PR DESCRIPTION
**Pull request by nehaoss**
**Description**
Enhances the eval create skill ([SKILL.md]) with comprehensive test case coverage patterns, multi-turn format support, and generation quality improvements — aligned with the PM CoCreate eval triage spec.

**Eval generation infrastructure**

1. Detect existing sets — skill now scans evaluations/ for existing test sets before asking the user, shows status table with counts, and offers to generate only missing categories or regenerate
2. Per-category grader threshold table — 0.7 standard, 0.5 for ambiguous-only; all single-response categories require both GeneralQualityGrader + CompareMeaningGrader
3. Evaluation set size limit section — max 100 cases per set, split into overflow folders with numeric suffix (e.g., topic-triggering-2/)

**TopicTriggering rewrite**

1. Rewrote TopicTriggering rules — pick 1-2 representative trigger queries per topic (no generated paraphrases), skip raw placeholders, target 30-60 positive tests
2. Added positive / boundary / negative distribution — each user-facing topic generates 3-5 test cases (≥1 positive, ≥1 boundary, ≥1 negative)
3. Boundary case types: typos/misspellings, casual abbreviations, synonym variants, very short input
4. Negative case types: out-of-scope, cross-domain mixing, privacy boundary, write-on-read-only, multi-intent confusion

**Extended categories with boundary + negative variants**

1. IntegrationData — added boundary (typo/synonym variants) and negative (privacy boundary, write-on-read-only) cases
2. AmbiguousTopic — added boundary (typos in ambiguous prompts) and negative (off-domain prompts that shouldn't trigger clarification) cases
3. MultiTurn — added 1-2 boundary conversations (typos mid-conversation) and 1-2 negative conversations (out-of-scope pivot mid-conversation)
4. RAI, SensitiveTopic, EQTopic unchanged (inherently single-type categories)

**Multi-turn format**

1. Documented correct kind: MultiTurnEvaluationCase with activities: format (not EvaluationData with rows:)
2. Parent uses GeneralQualityGrader only (no CompareMeaningGrader — it's single-response only)
3. Added Copilot Studio constraints (20 conversations, 6 question-response pairs, 500 chars per question)

**Expected response quality rules**

1. Based on empirical testing — improved compare-meaning scores
2. Never mention backend system names in expected responses (semantic mismatch penalty)
3. Keep assertions focused on observable user-facing behavior (not implementation details)
4. Use exact fallback message for unsupported topics (100% match)

**Related issue**
https://dev.azure.com/O365Exchange/O365%20Core/_workitems/edit/7414355
https://dev.azure.com/O365Exchange/O365%20Core/_workitems/edit/7395114
https://dev.azure.com/O365Exchange/O365%20Core/_workitems/edit/7395107

**Testing**

- Generated eval test sets using the updated skill against a live agent (FRE_ESS_IT_SPO_SF_m4.1) with 20+ topics
- Verified multi-turn files render as "Conversational chat (preview)" in Copilot Studio (not "Single response")
- Pushed 5 multi-turn test cases + 7 single-response eval sets (191 total test cases) to Copilot Studio successfully
- Confirmed CompareMeaningGrader scores improved with the response quality rules

image

### CODE REVIEW

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 1]
```diff
@@ -123,26 +205,152 @@ extensionData:
...
+
+Copilot Studio enforces a **maximum of 100 test cases per evaluation set**. If a
+category generates more than 100 child test cases, **split it into multiple sets**:
+
+1. Keep the first 100 tests in the original folder (e.g., `topic-triggering/`).
+2. Create additional folders with a numeric suffix (e.g., `topic-triggering-2/`,
+   `topic-triggering-3/`) for the overflow.
+3. Each overflow folder gets its own parent EvaluationSet file with a matching
+   `displayName` suffix (e.g., `"Topic Triggering 2"`).
+4. All split sets use the same grader configuration as the original.
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:222]: Dismissed — evals are stored in per-category folders (confirmed: evaluations/topic-triggering/, evaluations/ambiguous-prompts/, etc.). The folder-based split instruction is correct. The naming convention section earlier in the doc is the one that's slightly ambiguous — it doesn't explicitly mention the folder structure — but the split logic here matches reality.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 2]
```diff
@@ -66,9 +86,37 @@ Present the user with options. Keep it simple — one question:
...
+> | Responsible AI | {✅ exists (X tests) / ❌ missing} |
+> | Sensitive Topics | {✅ exists (X tests) / ❌ missing} |
+> | Emotional Intelligence | {✅ exists (X tests) / ❌ missing} |
+> | Integration Data | {✅ exists (X tests) / ❌ missing} |
+>
+> I can generate the **{M} missing** category/categories ({list missing names}),
+> or regenerate everything from scratch. What would you prefer?
+
+For the test count, count the number of child `.mcs.yml` files in each existing
+category folder (excluding the parent EvaluationSet file which shares the folder name).
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:107]: Dismissed — same as above. Evals use per-category folders, and the parent EvaluationSet file does share the folder name (e.g., ambiguous-prompts/ambiguous-prompts.mcs.yml). The count instruction is correct as written.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 3]
```diff
@@ -66,9 +86,37 @@ Present the user with options. Keep it simple — one question:
...
+>
+> | Category | Status |
+> |----------|--------|
+> | Topic Triggering | {✅ exists (X tests) / ❌ missing} |
+> | Ambiguous Prompts | {✅ exists (X tests) / ❌ missing} |
+> | Responsible AI | {✅ exists (X tests) / ❌ missing} |
+> | Sensitive Topics | {✅ exists (X tests) / ❌ missing} |
+> | Emotional Intelligence | {✅ exists (X tests) / ❌ missing} |
+> | Integration Data | {✅ exists (X tests) / ❌ missing} |
+>
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:102]: Good catch — fixed. Added Multi-Turn row to the status table.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 4]
```diff
@@ -47,9 +47,29 @@ via Dataverse. Test cases are stored as `botcomponent` records with
...
+| Topic Triggering | `topic-triggering` |
+| Ambiguous Prompts | `ambiguous-prompts` |
+| Responsible AI | `responsible-ai` |
+| Sensitive Topics | `sensitive-topics` |
+| Emotional Intelligence | `emotional-intelligence` |
+| Integration Data | `integration-data` |
+| Multi-Turn | `multi-turn` |
+
+For each category, check if a directory with that name exists under `evaluations/`
+AND contains at least one `.mcs.yml` file. Mark it as **existing** or **missing**.
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:68]: Dismissed — evals are stored in per-category folders (e.g., evaluations/topic-triggering/topic-triggering.mcs.yml). The directory check is correct. The naming convention section earlier in the doc could be clearer about the folder layout, but the scan logic here matches the actual structure.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 5]
```diff
@@ -47,9 +47,29 @@ via Dataverse. Test cases are stored as `botcomponent` records with
...
+## Step 2: Detect Existing Sets and Ask User About Scope
 
-Present the user with options. Keep it simple — one question:
+### 2a. Scan for existing evaluation sets
+
+Before asking the user what to generate, scan `{agent.folder}/evaluations/` for
+existing EvaluationSet parent files. The known categories and their folder names are:
+
+| Category | Folder name |
+|----------|-------------|
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:58]: Dismissed — these are actual folder names, not filename prefixes. The extracted agent structure uses evaluations/topic-triggering/, evaluations/ambiguous-prompts/, etc. "Folder name" is correct.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 6]
```diff
@@ -172,16 +381,127 @@ Generate one EvaluationSet file + child EvaluationData files for each applicable
...
+Multi-turn tests use a **different YAML kind** from single-response tests. Copilot
+Studio evaluates multi-turn tests as full conversations — each test case contains
+multiple question-response pairs that run sequentially in a single session. They
+appear in Copilot Studio under the **"Conversational chat (preview)"** data type.
+
+**CRITICAL**: Children MUST use `kind: MultiTurnEvaluationCase` (NOT `EvaluationData`).
+Using `EvaluationData` with multiple rows will create single-response tests that
+only show the first turn.
+
+**Format reference**: See `src/examples/ess-samples/ESSEvaluationSamples/EvalConversationTemplate.csv`
```
Comment (GrahamMcMynn) [inline solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: **Broken exemplar reference.** `src/examples/ess-samples/ESSEvaluationSamples/EvalConversationTemplate.csv` does not exist on this branch — neither `StarterTestSets/` nor `TemplatedTestSets/` contains it, and a `microsoft` org-wide code search returns zero hits. Because this skill is read by an LLM that will follow the reference, a broken link is worse than no link: the agent will either error or hallucinate a template format.

Please either (a) add the file under `ESSEvaluationSamples/`, (b) point to an existing multi-turn exemplar, or (c) drop the "Format reference" line and rely on the YAML example shown immediately below.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 7]
```diff
@@ -172,16 +381,127 @@ Generate one EvaluationSet file + child EvaluationData files for each applicable
...
+     — `expectedOutput`: the agent explains it cannot modify this data
 
 #### GeneralKnowledge
 
 **Only generate if** the user confirmed they have knowledge sources.
 
 1. Generate 10-15 general questions relevant to the agent's domain
-2. Omit `expectedOutput` — use the `GeneralQualityGrader` auto-grading
+2. Include `expectedOutput` describing the key information the agent should provide
+3. Use the exemplar `GeneralKnowledge-IT.csv` or `GeneralKnowledge-HR.csv` as reference for response style
```
Comment (GrahamMcMynn) [inline solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:414]: **GeneralKnowledge grader rule is half-wired.** This PR flips GeneralKnowledge from "omit expectedOutput, GeneralQualityGrader only" to "include expectedOutput and both graders" (lines 136-139, 172-174), but doesn't finish the change:

- The per-category threshold table at lines 143-150 omits GeneralKnowledge — only six categories are listed.
- This section (lines 399-405) gives no grader or threshold guidance at all.

An agent generating GeneralKnowledge eval sets has no documented threshold. Defaulting to 0.7 is likely wrong for free-form knowledge questions where wording varies widely across knowledge-base sources — the same reasoning that justified 0.5 for Ambiguous Prompts.

Please add a "General Knowledge" row to the threshold table (suggest 0.5 with that rationale) and add a step under `#### GeneralKnowledge` that restates: include both graders with the chosen threshold.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 8]
```diff
@@ -123,26 +206,152 @@ extensionData:
...
 
 Generate one EvaluationSet file + child EvaluationData files for each applicable category:
 
 #### TopicTriggering
 
-**For each user-facing topic:**
-1. Include all existing `triggerQueries` from the topic YAML as test prompts
-2. Generate 2-3 paraphrased variants (different wording, same intent)
-3. Set `expectedOutput` to a semantic description of what the topic should do — derive from:
+TopicTriggering is the only category that uses the **positive / boundary / negative**
```
Comment (GrahamMcMynn) [inline solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: **Intro claim contradicts the rest of this PR.** This says TopicTriggering is "the only category that uses the positive / boundary / negative distribution pattern", but the same PR adds positive/boundary/negative cases to:

- **AmbiguousTopic** (later in this file): "Positive (ambiguous) / Boundary / Negative"
- **IntegrationData**: "Positive / Boundary / Negative"
- **MultiTurn**: includes 1-2 boundary and 1-2 negative conversations

What is actually unique to TopicTriggering is the **strict minimum per topic** (≥1 of each type). Other categories use a flexible mix.

An LLM reading this intro literally may omit boundary and negative cases from those other categories, undoing the coverage gains this PR is trying to add. Suggest rewording to:

> TopicTriggering is the only category that enforces a strict minimum per topic (≥1 positive, ≥1 boundary, ≥1 negative). Other categories below — AmbiguousTopic, IntegrationData, MultiTurn — use a flexible mix of the same three variant types. RAI, SensitiveTopic, and EQTopic are single-type categories and do not use this pattern.

**Review by GrahamMcMynn — COMMENTED** [review]
Code review by ESS Code Review Agent (audit mode, posted with reviewer approval).

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 6 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: thank you graham, fixed this

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 7 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:414]: resolved this with 0.7 threshold for GeneralKnowledge

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 6 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: resolved, thank you

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 8 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: fixed this, thank you!!

**Review by GrahamMcMynn — APPROVED** [review]
Looks good after fixing those comments
